File: frontend/focus.js
/** Daily Scheduler's one-line Now / Next status. Empty when nothing is on now or later today. */
function nowNextLine(result, minute) {
  const parts = [];
  if (result.current) {
    const end = parseStart(result.current.start) + result.current.duration_min;
    parts.push("Now: " + result.current.title + " · " + formatDuration(end - minute) + " left");
  }
  if (result.next) {
    const wait = parseStart(result.next.start) - minute;
    parts.push("Next: " + result.next.title + " at " + result.next.start + (result.current ? "" : " (in " + formatDuration(wait) + ")"));
  }
  return parts.join("  →  ");
}

// Quick focus needs no task, so the section is there whenever someone is signed in.
function syncFocusSection() {
  document.getElementById("focus-section").hidden = !account;
}

function updateLiveDisplay(nowDate) {
  const target = document.getElementById("now-next");
  if (!target) return;
  const now = nowDate || new Date();
  const info = currentDateInfo(now);
  const state = weeks.get(info.week);
  const status = nowNextLine(nowAndNext(scheduledBlocksForState(state), info.day, info.minute), info.minute);
  target.textContent = status;
  target.hidden = !status;
  document.querySelectorAll(".current-time-line").forEach(function (line) {
    if (typeof line.remove === "function") line.remove();
    else line.hidden = true;
  });
  if (selectedWeek !== info.week || info.minute < DAY_START_MIN || info.minute >= DAY_END_MIN) return;
  const lane = Array.from(weekEl.children).find(function (child) {
    return child.classList && child.classList.contains("day-lane") && Number(child.dataset.day) === info.day;
  });
  if (!lane) return;
  const line = document.createElement("div");
  line.className = "current-time-line";
  line.style.top = ((info.minute - DAY_START_MIN) / 60) * hourHeightRem() + "rem";
  line.title = "Current time " + formatMinute(info.minute);
  lane.appendChild(line);
}

function phaseDurationMs(phase) {
  const minutes = phase === "work" ? prefs.timer_work_min :
    (phase === "long_break" ? prefs.timer_long_break_min : prefs.timer_break_min);
  return minutes * 60000;
}

function formatCountdown(milliseconds) {
  const seconds = Math.max(0, Math.ceil(milliseconds / 1000));
  return String(Math.floor(seconds / 60)).padStart(2, "0") + ":" + String(seconds % 60).padStart(2, "0");
}

const MAX_ESTIMATE_MIN = 7140;
const MORE_TIME_CHOICES = [15, 30, 45, 60, 90, 120, 180, 240];
const FOCUS_STORAGE_PREFIX = "flexweek.focus.";
const FOCUS_PHASES = ["work", "break", "long_break", "ended"];
const FOCUS_PHASE_LABEL = { work: "Focus session", break: "Break", long_break: "Long break", ended: "Session done" };
// After a homework session ends, whether the student is choosing how much more time it needs.
let focusAskingMore = false;

function renderFocusPanel(nowMs) {
  const panel = document.getElementById("focus-panel");
  if (!panel) return;
  panel.hidden = !focusState;
  syncFocusSection();
  if (!focusState) return;
  const ended = focusState.phase === "ended";
  document.getElementById("focus-task").textContent = focusState.title;
  document.getElementById("focus-phase").textContent = FOCUS_PHASE_LABEL[focusState.phase];
  const left = focusState.running ? focusState.endsAt - (nowMs || Date.now()) : focusState.remainingMs;
  document.getElementById("focus-time").textContent = formatCountdown(left);
  document.getElementById("focus-pause").textContent = focusState.running ? "Pause" : "Resume";
  document.getElementById("focus-controls").hidden = ended;
  document.getElementById("focus-choices").hidden = !ended || focusAskingMore;
  document.getElementById("focus-more-form").hidden = !ended || !focusAskingMore;
}

function resetFocusTimer(hide) {
  if (focusTimer) clearInterval(focusTimer);
  focusTimer = null;
  focusState = null;
  focusAskingMore = false;
  const panel = document.getElementById("focus-panel");
  if (panel && hide !== false) panel.hidden = true;
  syncFocusSection();
  persistFocus();
}

/** The timer outlives a reload in sessionStorage, as ids and times only; titles come back after sign-in. */
function persistFocus() {
  if (!account) return;
  try {
    const key = FOCUS_STORAGE_PREFIX + account.id;
    if (!focusState) {
      sessionStorage.removeItem(key);
      return;
    }
    sessionStorage.setItem(key, JSON.stringify({
      assignmentId: focusState.assignmentId, sessionId: focusState.blockId, weekStart: focusState.weekStart,
      day: focusState.day, start: focusState.start, phase: focusState.phase, cycles: focusState.cycles,
      endsAt: focusState.running ? focusState.endsAt : null,
      remainingMs: focusState.running ? null : focusState.remainingMs,
    }));
  } catch { /* Without sessionStorage the timer still runs until the page closes. */ }
}

function forgetFocus(accountId) {
  try { sessionStorage.removeItem(FOCUS_STORAGE_PREFIX + accountId); } catch { /* Nothing was stored. */ }
}

/** After sign-in, resume this account's timer from before a reload, and drop any other account's. */
function restoreFocus() {
  let saved = null;
  try {
    const mine = FOCUS_STORAGE_PREFIX + account.id;
    for (let index = sessionStorage.length - 1; index >= 0; index -= 1) {
      const key = sessionStorage.key(index);
      if (key && key.startsWith(FOCUS_STORAGE_PREFIX) && key !== mine) sessionStorage.removeItem(key);
    }
    saved = JSON.parse(sessionStorage.getItem(mine) || "null");
  } catch { return; }
  if (focusState || !saved) return;
  const assignment = typeof saved.assignmentId === "string" ? assignments.get(saved.assignmentId) : null;
  const valid = FOCUS_PHASES.includes(saved.phase) && isWeekStart(saved.weekStart) && Number.isInteger(saved.cycles) &&
    (typeof saved.endsAt === "number" || typeof saved.remainingMs === "number");
  // Homework finished or deleted since then has nothing left to time.
  if (!valid || (saved.assignmentId && (!assignment || assignment.completed))) {
    forgetFocus(account.id);
    return;
  }
  const sessionId = typeof saved.sessionId === "string" ? saved.sessionId : null;
  const week = weeks.get(saved.weekStart);
  const block = sessionId && week ? week.blocks.find(function (item) { return item.id === sessionId; }) : null;
  focusState = {
    weekStart: saved.weekStart, blockId: sessionId, assignmentId: assignment ? assignment.id : null,
    day: Number.isInteger(saved.day) ? saved.day : null, start: typeof saved.start === "string" ? saved.start : null,
    title: sessionId ? (assignment || block || { title: "Focus session" }).title : "Quick focus",
    phase: saved.phase, cycles: saved.cycles, running: typeof saved.endsAt === "number",
    endsAt: saved.endsAt || 0, remainingMs: saved.remainingMs || 0,
  };
  startFocusInterval();
  if (focusState.running && focusState.endsAt <= Date.now()) {
    focusState.running = false;
    focusState.remainingMs = 0;
    if (focusState.phase === "work") {
      // It ran out while the page was closed: count it now and ask what comes next.
      advanceFocusPhase(true);
    } else {
      focusState.phase = "work";
      focusState.remainingMs = phaseDurationMs("work");
      setStatus("The break ended while FlexWeek was closed. Press Resume to focus again.");
    }
  }
  renderFocusPanel();
  persistFocus();
}

function resolveFocusPlacement(blockId, day) {
  const state = weekState();
  const source = state.blocks.find(function (block) { return block.id === blockId; });
  if (!source) return null;
  const placed = source.start ? source : state.trace && state.trace.placed.find(function (block) {
    return block.id === blockId && (day === undefined || block.days.indexOf(day) !== -1);
  });
  return placed && placed.start ? { source: source, placed: placed, day: placed.days[0] } : null;
}

function startFocus(blockId, day) {
  if (!account || saving || focusBusy) return false;
  const found = resolveFocusPlacement(blockId, day);
  if (!found || found.source.completed || found.source.pomodoro_role === "break") {
    setStatus("Place an unfinished work block before starting focus.");
    return false;
  }
  return beginFocus({
    weekStart: selectedWeek, blockId: blockId, assignmentId: found.source.assignment_id || null,
    day: found.day, start: found.placed.start, title: found.source.title,
  });
}

/** Quick focus times work that is not on the calendar, so it credits nothing. */
function startQuickFocus() {
  if (!account || saving || focusBusy) return false;
  return beginFocus({ weekStart: selectedWeek, blockId: null, assignmentId: null, day: null, start: null, title: "Quick focus" });
}

function beginFocus(target) {
  if (focusState && !confirm("Stop the timer for " + focusState.title + " and start " + target.title + " instead?")) {
    return false;
  }
  resetFocusTimer();
  focusState = {
    ...target, phase: "work", cycles: 0, running: true,
    remainingMs: phaseDurationMs("work"), endsAt: Date.now() + phaseDurationMs("work"),
  };
  startFocusInterval();
  renderFocusPanel();
  persistFocus();
  return true;
}

function startFocusInterval() {
  if (focusTimer) clearInterval(focusTimer);
  focusTimer = setInterval(function () { focusTick(); }, 500);
  if (focusTimer && typeof focusTimer.unref === "function") focusTimer.unref();
}

async function creditFocusSession() {
  // Quick focus has no task, so it credits nothing.
  if (!focusState || !focusState.blockId) return;
  if (focusState.assignmentId) {
    // Minutes count toward the assignment and never finish anything; that is the student's choice.
    const assignment = assignments.get(focusState.assignmentId);
    // A session's own focus fields must stay 0, so without its assignment there is nowhere to credit.
    if (!assignment) {
      setStatus("This homework did not load, so the focus time was not counted. Reload the week and try again.");
      return;
    }
    putAssignment({
      ...assignment,
      focus_sessions: Math.min(9999, (assignment.focus_sessions || 0) + 1),
      focus_minutes: Math.min(71400, (assignment.focus_minutes || 0) + prefs.timer_work_min),
    });
  } else {
    if (focusState.weekStart !== selectedWeek) return;
    const block = weekState().blocks.find(function (item) { return item.id === focusState.blockId; });
    if (!block) return;
    block.focus_sessions = Math.min(9999, (block.focus_sessions || 0) + 1);
    block.focus_minutes = Math.min(71400, (block.focus_minutes || 0) + prefs.timer_work_min);
  }
  // Focus credit is progress, not an edit, so Undo leaves it alone.
  absorbIntoHistory();
  await saveWeek();
  renderWeek();
}

async function advanceFocusPhase(completed) {
  if (!focusState || focusBusy || focusState.phase === "ended") return;
  focusBusy = true;
  try {
    if (focusState.phase === "work") {
      if (completed) await creditFocusSession();
      if (!focusState) return;
      focusState.cycles += completed ? 1 : 0;
      if (completed && focusState.assignmentId && assignments.has(focusState.assignmentId)) {
        // Homework asks what comes next instead of rolling into a break.
        focusState.phase = "ended";
        focusState.running = false;
        focusState.remainingMs = 0;
        maybeNotify("Focus session done", focusState.title, true, "soft");
        renderFocusPanel();
        persistFocus();
        return;
      }
      beginFocusBreak();
    } else {
      setFocusPhase("work");
    }
  } finally {
    focusBusy = false;
  }
}

function setFocusPhase(phase) {
  focusAskingMore = false;
  focusState.phase = phase;
  focusState.running = true;
  focusState.remainingMs = phaseDurationMs(phase);
  focusState.endsAt = Date.now() + focusState.remainingMs;
  maybeNotify(FOCUS_PHASE_LABEL[phase], focusState.title, true, phase === "work" ? "bright" : "soft");
  renderFocusPanel();
  persistFocus();
}

function beginFocusBreak() {
  setFocusPhase(focusState.cycles > 0 && focusState.cycles % prefs.timer_long_break_every === 0 ? "long_break" : "break");
}

/** Finished: the homework is done and the session keeps the slot it was worked in, saved together. */
async function finishFocusedHomework() {
  if (!focusState || focusState.phase !== "ended" || saving || focusBusy) return false;
  const target = focusState;
  if (selectedWeek !== target.weekStart && !await selectWeek(target.weekStart)) return false;
  const assignment = assignments.get(target.assignmentId);
  if (!assignment || focusState !== target) return false;
  const block = weekState().blocks.find(function (item) { return item.id === target.blockId; });
  if (block) {
    block.completed = true;
    if (block.kind === "flexible") {
      if (target.start && Number.isInteger(target.day) && block.days.includes(target.day)) {
        block.start = target.start;
        block.completed_day = target.day;
      } else if (!Number.isInteger(block.completed_day)) {
        block.start = null;
      }
    }
  }
  putAssignment({ ...assignment, completed: true, completed_at: localStamp() });
  resetFocusTimer();
  commitWeek("Finished " + assignment.title + ".", "finishing " + assignment.title);
  return true;
}

function moreTimeChoices(assignment) {
  return MORE_TIME_CHOICES.filter(function (minutes) { return assignment.estimate_min + minutes <= MAX_ESTIMATE_MIN; });
}

/** Need more time: add minutes on the 15-minute grid to the homework's total, then take the break. */
async function addFocusTime(minutes) {
  if (!focusState || focusState.phase !== "ended" || saving || focusBusy) return false;
  const assignment = assignments.get(focusState.assignmentId);
  if (!assignment || !moreTimeChoices(assignment).includes(minutes)) {
    setStatus("Choose how much more time it needs.");
    return false;
  }
  putAssignment({ ...assignment, estimate_min: assignment.estimate_min + minutes });
  recordStep("adding time to " + assignment.title);
  beginFocusBreak();
  const saved = await saveWeek();
  renderWeek();
  if (saved) {
    setStatus("Added " + formatDuration(minutes) + " to " + assignment.title + ". Plan it under Continuing, then press Solve.");
  }
  return saved;
}

function takeFocusBreak() {
  if (!focusState || focusState.phase !== "ended" || focusBusy) return false;
  beginFocusBreak();
  return true;
}

function focusTick(nowMs) {
  if (!focusState || !focusState.running) return;
  const now = nowMs || Date.now();
  if (focusState.endsAt <= now) {
    focusState.running = false;
    focusState.remainingMs = 0;
    advanceFocusPhase(true);
  }
  renderFocusPanel(now);
}

function toggleFocusPause() {
  if (!focusState || focusBusy || saving || focusState.phase === "ended") return;
  if (focusState.running) {
    focusState.remainingMs = Math.max(0, focusState.endsAt - Date.now());
    focusState.running = false;
  } else {
    focusState.running = true;
    focusState.endsAt = Date.now() + focusState.remainingMs;
  }
  renderFocusPanel();
  persistFocus();
}

function renderFocusTasks() {
  const list = document.getElementById("focus-tasks");
  if (!list) return;
  list.replaceChildren();
  const scheduled = scheduledBlocksForState(weekState());
  const placementById = new Map(scheduled.map(function (block) { return [block.id, block]; }));
  // Only work that has a time and is not done can be focused; school and breaks cannot.
  weekState().blocks.filter(function (block) {
    return (block.kind === "flexible" || block.pomodoro_role === "work") && !block.completed &&
      !(assignmentOf(block) || {}).completed && placementById.has(block.id);
  }).forEach(function (block) {
    const placement = placementById.get(block.id);
    const progress = assignmentOf(block) || block;
    const item = document.createElement("li");
    const name = document.createElement("strong");
    name.textContent = block.title;
    const count = document.createElement("small");
    count.textContent = DAYS[placement.days[0]] + " " + placement.start + (progress.focus_sessions
      ? " · " + progress.focus_sessions + (progress.focus_sessions === 1 ? " session, " : " sessions, ") + formatDuration(progress.focus_minutes || 0)
      : "");
    const start = document.createElement("button");
    start.type = "button";
    start.textContent = "Start";
    start.addEventListener("click", function () {
      const placed = placementById.get(block.id);
      startFocus(block.id, placed && placed.days[0]);
    });
    item.appendChild(name);
    item.appendChild(count);
    item.appendChild(start);
    list.appendChild(item);
  });
  syncFocusSection();
}

document.getElementById("focus-pause").addEventListener("click", toggleFocusPause);
document.getElementById("focus-skip").addEventListener("click", function () {
  if (saving || focusBusy) return;
  advanceFocusPhase(false);
});
document.getElementById("focus-reset").addEventListener("click", function () { resetFocusTimer(); });
document.getElementById("focus-quick").addEventListener("click", function () { startQuickFocus(); });
document.getElementById("focus-finished").addEventListener("click", function () { finishFocusedHomework(); });
document.getElementById("focus-break").addEventListener("click", function () { takeFocusBreak(); });
document.getElementById("focus-more").addEventListener("click", function () {
  const assignment = focusState && assignments.get(focusState.assignmentId);
  const choices = assignment ? moreTimeChoices(assignment) : [];
  if (!choices.length) {
    setStatus("This homework already has the most time FlexWeek allows.");
    return;
  }
  const select = document.getElementById("focus-more-min");
  fillOptions(select, choices.map(function (minutes) { return [minutes, formatDuration(minutes)]; }));
  select.value = String(choices[Math.min(1, choices.length - 1)]);
  focusAskingMore = true;
  renderFocusPanel();
});
document.getElementById("focus-more-add").addEventListener("click", function () {
  addFocusTime(Number(document.getElementById("focus-more-min").value));
});
document.getElementById("focus-more-cancel").addEventListener("click", function () {
  focusAskingMore = false;
  renderFocusPanel();
});

File: desktop/native/focus.py
"""Focus timer math. Wall-clock phases, persist payloads and credit rules. No Qt."""

from __future__ import annotations

from copy import deepcopy

from backend.slots import hhmm_to_minutes
from backend.weeks import is_week_start
from desktop.native.reuse import occurrence_days
from desktop.native.weekmodel import length_label

FOCUS_PHASES = ("work", "break", "long_break", "ended")
FOCUS_PHASE_LABEL = {
    "work": "Focus session",
    "break": "Break",
    "long_break": "Long break",
    "ended": "Session finished",
}
MAX_ESTIMATE_MIN = 7140
MAX_FOCUS_MINUTES = 71400
MAX_FOCUS_SESSIONS = 9999
MORE_TIME_CHOICES = (15, 30, 45, 60, 90, 120, 180, 240)
DEFAULT_TIMERS = {
    "timer_work_min": 30,
    "timer_break_min": 15,
    "timer_long_break_min": 30,
    "timer_long_break_every": 4,
}


def phase_duration_ms(phase: str, prefs: dict | None) -> int:
    source = prefs or {}
    if phase == "work":
        minutes = int(source.get("timer_work_min") or DEFAULT_TIMERS["timer_work_min"])
    elif phase == "long_break":
        minutes = int(source.get("timer_long_break_min") or DEFAULT_TIMERS["timer_long_break_min"])
    else:
        minutes = int(source.get("timer_break_min") or DEFAULT_TIMERS["timer_break_min"])
    return minutes * 60_000


def format_countdown(milliseconds: int) -> str:
    seconds = max(0, (milliseconds + 999) // 1000)
    return f"{seconds // 60:02d}:{seconds % 60:02d}"


def remaining_ms(state: dict, now_ms: int) -> int:
    if state.get("running"):
        return max(0, int(state.get("endsAt") or 0) - now_ms)
    return max(0, int(state.get("remainingMs") or 0))


def more_time_choices(estimate_min: int) -> list[int]:
    return [minutes for minutes in MORE_TIME_CHOICES if estimate_min + minutes <= MAX_ESTIMATE_MIN]


def persist_payload(state: dict | None) -> dict | None:
    if state is None:
        return None
    running = bool(state.get("running"))
    return {
        "assignmentId": state.get("assignmentId"),
        "sessionId": state.get("blockId"),
        "weekStart": state.get("weekStart"),
        "day": state.get("day"),
        "start": state.get("start"),
        "phase": state.get("phase"),
        "cycles": int(state.get("cycles") or 0),
        "endsAt": state.get("endsAt") if running else None,
        "remainingMs": None if running else state.get("remainingMs"),
    }


def restore_state(saved: dict | None, *, assignments: dict, blocks: list[dict], now_ms: int) -> dict | None:
    if not isinstance(saved, dict):
        return None
    phase = saved.get("phase")
    week_start = saved.get("weekStart")
    cycles = saved.get("cycles")
    if (
        phase not in FOCUS_PHASES
        or not isinstance(week_start, str)
        or not is_week_start(week_start)
        or not isinstance(cycles, int)
    ):
        return None
    has_end = isinstance(saved.get("endsAt"), (int, float))
    has_remaining = isinstance(saved.get("remainingMs"), (int, float))
    if not has_end and not has_remaining:
        return None
    assignment_id = saved.get("assignmentId") if isinstance(saved.get("assignmentId"), str) else None
    assignment = assignments.get(assignment_id) if assignment_id else None
    if assignment_id and (assignment is None or assignment.get("completed")):
        return None
    block_id = saved.get("sessionId") if isinstance(saved.get("sessionId"), str) else None
    block = next((item for item in blocks if item["id"] == block_id), None) if block_id else None
    running = isinstance(saved.get("endsAt"), (int, float))
    title = "Quick focus"
    if block_id:
        title = (assignment or block or {"title": "Focus session"})["title"]
    state = {
        "weekStart": week_start,
        "blockId": block_id,
        "assignmentId": assignment["id"] if assignment else None,
        "day": saved.get("day") if isinstance(saved.get("day"), int) else None,
        "start": saved.get("start") if isinstance(saved.get("start"), str) else None,
        "title": title,
        "phase": phase,
        "cycles": cycles,
        "running": running,
        "endsAt": int(saved.get("endsAt") or 0),
        "remainingMs": int(saved.get("remainingMs") or 0),
    }
    if running and state["endsAt"] <= now_ms:
        state["running"] = False
        state["remainingMs"] = 0
        state["expired"] = True
    return state


def begin_state(target: dict, prefs: dict | None, now_ms: int) -> dict:
    duration = phase_duration_ms("work", prefs)
    return {
        **target,
        "phase": "work",
        "cycles": 0,
        "running": True,
        "remainingMs": duration,
        "endsAt": now_ms + duration,
    }


def pause_state(state: dict, now_ms: int) -> dict:
    next_state = deepcopy(state)
    if next_state.get("phase") == "ended":
        return next_state
    if next_state.get("running"):
        next_state["remainingMs"] = remaining_ms(next_state, now_ms)
        next_state["running"] = False
    else:
        next_state["running"] = True
        next_state["endsAt"] = now_ms + int(next_state.get("remainingMs") or 0)
    return next_state


def set_phase(state: dict, phase: str, prefs: dict | None, now_ms: int) -> dict:
    duration = phase_duration_ms(phase, prefs)
    next_state = deepcopy(state)
    next_state["phase"] = phase
    next_state["running"] = True
    next_state["remainingMs"] = duration
    next_state["endsAt"] = now_ms + duration
    return next_state


def break_phase(cycles: int, prefs: dict | None) -> str:
    every = int((prefs or {}).get("timer_long_break_every") or DEFAULT_TIMERS["timer_long_break_every"])
    if cycles > 0 and cycles % every == 0:
        return "long_break"
    return "break"


def credit_target(state: dict, assignment: dict | None, block: dict | None, work_min: int) -> dict | None:
    """Return the mutated homework or block after one completed work phase. None if nothing to credit."""
    if not state.get("blockId"):
        return None
    amount = min(MAX_FOCUS_MINUTES, work_min)
    if state.get("assignmentId"):
        if assignment is None:
            return None
        updated = deepcopy(assignment)
        updated["focus_sessions"] = min(MAX_FOCUS_SESSIONS, int(updated.get("focus_sessions") or 0) + 1)
        updated["focus_minutes"] = min(MAX_FOCUS_MINUTES, int(updated.get("focus_minutes") or 0) + amount)
        return updated
    if block is None:
        return None
    updated = deepcopy(block)
    updated["focus_sessions"] = min(MAX_FOCUS_SESSIONS, int(updated.get("focus_sessions") or 0) + 1)
    updated["focus_minutes"] = min(MAX_FOCUS_MINUTES, int(updated.get("focus_minutes") or 0) + amount)
    return updated


def focus_candidates(blocks: list[dict], assignments: dict, trace: dict | None) -> list[dict]:
    placed = {item["id"]: item for item in (trace or {}).get("placed") or []}
    result = []
    for block in blocks:
        work = block.get("kind") == "flexible" or block.get("pomodoro_role") == "work"
        if not work or block.get("completed"):
            continue
        assignment = assignments.get(block.get("assignment_id"))
        if assignment and assignment.get("completed"):
            continue
        placement = placed.get(block["id"], block if block.get("start") else None)
        if placement is None or not placement.get("start"):
            continue
        result.append(
            {
                "id": block["id"],
                "title": block["title"],
                "day": (placement.get("days") or [None])[0],
                "start": placement["start"],
                "focus_sessions": int((assignment or block).get("focus_sessions") or 0),
                "focus_minutes": int((assignment or block).get("focus_minutes") or 0),
            }
        )
    return result


def now_and_next(blocks: list[dict], day: int, minute: int) -> dict:
    active = []
    for block in blocks or []:
        if not block.get("start") or block.get("completed"):
            continue
        if day in (block.get("missed_days") or []):
            continue
        if day not in occurrence_days(block):
            continue
        active.append(block)
    active.sort(key=lambda item: hhmm_to_minutes(item["start"]))
    current = None
    following = None
    for block in active:
        start = hhmm_to_minutes(block["start"])
        end = start + int(block["duration_min"])
        if current is None and start <= minute < end:
            current = block
        if following is None and start > minute:
            following = block
    return {"current": current, "next": following}


def now_next_line(result: dict, minute: int) -> str:
    parts = []
    current = result.get("current")
    following = result.get("next")
    if current:
        end = hhmm_to_minutes(current["start"]) + int(current["duration_min"])
        parts.append(f"Now: {current['title']} · {length_label(end - minute)} left")
    if following:
        wait = hhmm_to_minutes(following["start"]) - minute
        suffix = "" if current else f" (in {length_label(wait)})"
        parts.append(f"Next: {following['title']} at {following['start']}{suffix}")
    return "  →  ".join(parts)
